### PR TITLE
DOC: Fix docstring and examples for rfn.get_names*

### LINKED ISSUE #12485: BUG: recfunctions and names
@eric-wieser suggested this may be an issue based on the doctest failure [noted here](https://github.com/numpy/numpy/pull/12253#discussion_r238081552)

```python
import numpy as np
from numpy.lib import recfunctions as rfn
rfn.get_names_flat(np.empty((1,), dtype=int)) is None
```

```
AttributeError: 'numpy.ndarray' object has no attribute 'names'
```

### PR DESCRIPTION
**Pull request by bsipocz**
This is to close #12485 

No point in getting names when they are not available for e.g scalar dtypes, and based on the arg description passing arrays instead of dtype objects is not a valid input indeed. So the issue only needed updates to the examples.

### CODE REVIEW

File: numpy/lib/recfunctions.py [hunk 1]
```diff
@@ -105,7 +105,8 @@ def _get_fieldspec(dtype):
 
 def get_names(adtype):
     """
-    Returns the field names of the input datatype as a tuple.
+    Returns the field names of the input datatype as a tuple. Input datatype
+    has to have fields otherwise error is raised.
```
Comment (mattip) [inline numpy/lib/recfunctions.py:109]:
```suggestion
    must have fields otherwise error is raised.
```

File: numpy/lib/recfunctions.py [hunk 2]
```diff
@@ -141,8 +137,9 @@ def get_names(adtype):
 
 def get_names_flat(adtype):
     """
-    Returns the field names of the input datatype as a tuple. Nested structure
-    are flattened beforehand.
+    Returns the field names of the input datatype as a tuple. Input datatype
+    has to have fields otherwise error is raised.
```
Comment (mattip) [inline numpy/lib/recfunctions.py:141]:
```suggestion
    must have fields otherwise error is raised.
```

**Comment by mattip** [comment]
Are the examples run in doctests? LGTM, +/- one small ""nit"

**Comment by bsipocz** [comment]
> Are the examples run in doctests? 

Hmm, I assumed they are, but now I'm not sure as I cannot make this file fail locally with bogus output lines. Probably I just don't know the right permutation of options for runtests.py, so far nothing seems to pick it up, and running pytest directly doesn't seem to work either.
otoh `numpy.test(doctest=True)`  raises an error, and I don't find any more trace in the docs.

**Comment by bsipocz** [comment]
OK, so running `rundocs('numpy/lib/recfunctions.py')` passes (and I double-checked that the function works as expected). However, it's in `_private` so I'm a bit confused about its status and the reasons it's being decommissioned.

**Comment by mattip** [comment]
Thanks @bsipocz

**Pull request merged**